spacy/kb/candidate.pyx

```diff
@@ -1,74 +1,116 @@
 # cython: infer_types=True, profile=True
 
-from typing import Iterable
+from ..typedefs cimport hash_t
+
 from .kb cimport KnowledgeBase
-from ..tokens import Span
+from .kb_in_memory cimport InMemoryLookupKB
+from ..errors import Errors
 
 cdef class Candidate:
-    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
-    to a specific `entity` from a Knowledge Base. This will be used as input for the entity linking
+    """A `Candidate` object refers to a textual mention that may or may not be resolved
+    to a specific `entity_id` from a Knowledge Base. This will be used as input for the entity_id linking
     algorithm which will disambiguate the various candidates to the correct one.
-    Each candidate (alias, entity) pair is assigned a certain prior probability.
+    Each candidate (mention, entity_id) pair is assigned a certain prior probability.
 
     DOCS: https://spacy.io/api/kb/#candidate-init
     """
 
-    def __init__(self, KnowledgeBase kb, entity_hash, entity_freq, entity_vector, alias_hash, prior_prob):
-        self.kb = kb
-        self.entity_hash = entity_hash
-        self.entity_freq = entity_freq
-        self.entity_vector = entity_vector
-        self.alias_hash = alias_hash
-        self.prior_prob = prior_prob
+    def __init__(
+        self,
+        entity_id: str,
+        entity_vector: vector[float],
+        prior_prob: float,
+    ):
```

Comment:
The alias/mention is not here?

Reply:
Yes, because it should be up to the KB and `Candidate` child class implementations how to store and retrieve mentions. Originally `Candidate` stored `mention` as string, but this meant that `InMemoryLookupKB` had an unnecessarily large memory footprint. Instead `Candidate` only implements the (theoretically abstract :cry:) property to return `mention`.

E. g. for `WikiKB` I wouldn't want to look up mentions every time, since that would mean querying the DB on disk for each individual lookup. Instead I'd store `mention` in my candidate, which can be an implementation detail of `WikiKBCandidate`.

Reply:
Hm. It just feels weird to me have a constructor of an abstract class that can never be used as such. Because semantically it makes no sense to have an entity ID and a prior probability without actually also having a mention?

Reply:
And so it feels like this constructor here doesn't really explain to users what a `Candidate` is / should be.

Reply:
> Hm. It just feels weird to me have a constructor of an abstract class that can never be used as such.

I'd argue that it can be used as a base constructor called from the inheriting class. And an abstract class can never be used as such. Having said that, I agree that this is a not-quite-here-not-quite-there situation. Ideally I'd have the mention in here as well, but then we burden `InMemoryCandidate` with storing this too.

Reply:
Alternative: we change `Candidate` to be purely functional, i. e. an interface, not storing _any_ attributes. This would at least be more consistent.

Reply:
And the only place candidates are instantiated are in the knowledge base classes, so there is no need for a unified `__init__()` signature.

Reply:
Yes, I was arriving to the same conclusion :-) https://github.com/explosion/spaCy/pull/12341/files#r1132433003